### PR TITLE
[Super Cache] Prevent deleting freshly preloaded files when GC time set to 0

### PR DESCRIPTION
**Pull request by thingalon**
Fixes https://wordpress.org/support/topic/preload-mode-skips-caching-pages-or-deletes-already-cached-pages

When preloads finish, `prune_super_cache` is called to clean out leftovers. Unfortunately, if `$cache_max_time` is set to 0, preload will delete any files at least one second old - which frequently includes preloaded files. It causes intermittent post-preload disappearances.

This PR fixes this by ensuring the prune process will not delete files based on mtime if `$cache_max_time` is 0.

## Proposed changes:
* Prevent prune_super_cache from deleting files if `$cache_max_time` is set to 0.

### Other information:
- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

n/a

## Does this pull request change what data or activity we track or use?

no

## Testing instructions:
* Set cache timeout to 0
* Manually preload over and over
* Without this patch, sometimes files created near the start of the preload will get deleted at the end
* With this patch, all files created during the preload should remain. :)

### CODE REVIEW

File: projects/plugins/super-cache/wp-cache-phase2.php [hunk 1]
```diff
@@ -2661,6 +2661,12 @@ function prune_super_cache( $directory, $force = false, $rename = false ) {
 			wp_cache_debug( "gc: could not delete $directory as it's protected.", 2 );
 			$oktodelete = false;
 		}
+
+		if ( ! $rename && $cache_max_time < 1 ) {
```
Comment (pyronaur) [inline projects/plugins/super-cache/wp-cache-phase2.php]:
```suggestion
		if ( ! $rename && $cache_max_time === 0 ) {
```

**Review by pyronaur — COMMENTED** [review]
Maybe the code should be updated to match the debug message?

**Review by donnchawp — CHANGES_REQUESTED** [review]
Good catch. I wonder if it would be better to put the $cache_max_time check at the top of prune_super_cache instead? It would save traversing down the directories. But we'd have to check everywhere else we use prune_super_cache().

Or in wp_cron_preload_cache we set the global $cache_max_time variable to the current time minus the filemtime of "preload_mutex.tmp". That file is created when preloading starts. Set that before calling prune_super_cache() at line 3248 and preloaded files should be safe.

**Comment by thingalon** [comment]
> Good catch. I wonder if it would be better to put the $cache_max_time check at the top of prune_super_cache instead? It would save traversing down the directories. But we'd have to check everywhere else we use prune_super_cache().

I thought about doing this, but I wasn't sure about the `$rename` references I found throughout that code. I wasn't sure if there was any possibility of preloaded content getting renamed, and didn't want to risk causing new issues. 

If that's not a consideration, then I would advocate for this. :) 

> Or in wp_cron_preload_cache we set the global $cache_max_time variable to the current time minus the filemtime of "preload_mutex.tmp". That file is created when preloading starts. Set that before calling prune_super_cache() at line 3248 and preloaded files should be safe.

This is also a very clever idea. I like that this would cause the preload system to clean out irrelevant files if you (for example) deleted a category since the last refresh.

We'd want to add a buffer when calculating a `$cache_max_time` override in case the system clock ticks over a second between starting on the prune job, and examining a particular file for pruning.

However, this would mean that we'd delete any cached content that isn't preloaded. So cached search results with a GET param of `?s=` which may be fairly recent will get wiped. Is that a problem?

**Review by pyronaur — DISMISSED** [review]

**Comment by thingalon** [comment]
I made a couple of changes in b1532bb416 based on @donnchawp 's suggestions:
1. I found some pre-existing logic for setting a max cache time before calling prune in the preload code already - there were just some code paths which could reach the prune call before setting an appropriate max cache time. So I moved that logic up, and
2. I updated the prune method to exit early if max time is set to 0.

Based on my testing, this fixes the issue, and preload should still clean out unused files after a long time. :-)

**Comment by donnchawp** [comment]
I think we can close this. The mass deletion of files seemed to be caused in part by how slow that person's server was, but we've made changes elsewhere to speed up preload or at least make it more reliable on slow servers.

**Pull request closed without merging**